### ISSUE TITLE
Bug with MapForge in OpenStreetMapViewer

### ISSUE THREAD

**Issue by pavelkv96** [osmdroid/osmdroid#1145]
## Issue Type
Bug

## Description and/or steps/code to reproduce the problem
I have file ".map".
I open example More Sample -> TileProviders -> Mapsforge tiles.
## And I zooming by level 9, I seeing next... It's good.
![screenshot_1535450704](https://user-images.githubusercontent.com/29233766/44717013-8e268080-aac4-11e8-8355-b6b4bf62a179.png)

## When I set zoom in (10 level). I have next... It's good too.
![screenshot_1535450713](https://user-images.githubusercontent.com/29233766/44717165-fffeca00-aac4-11e8-8935-b241dc91bcf0.png)

## But when I return to level 9, I have next...
![screenshot_1535450719](https://user-images.githubusercontent.com/29233766/44717314-5bc95300-aac5-11e8-935b-c658699f0715.png)

### Version of osmdroid the issue relates to:

6.0.2

**Comment by monsieurtanuki** [comment]
@pavelkv96 My questions:
* does it happen with other tile providers than your .map files?
* are you 100% sure you're on _integer_ zoom level 9 when you go back to zoom 9 (hint: display the zoom level in the log using `MapView.getZoomLevelDouble()` in order to be sure). If it's zoom level 8.9999 that could be an explanation, for the missing title and the larger gray zone on the bottom of the screen
* I don't know anything about MapForge .map files - are there expiration dates for their tiles?

**Comment by pavelkv96** [comment]
@monsieurtanuki Sorry.
I did not notice gray zone on the bottom, when I did screenshot.
I would like to show an error with the display of the names of settlements.
When I do zoom in and I do next zoom out. I get a part of the city name for this level

**Comment by monsieurtanuki** [comment]
@pavelkv96 Thanks for the .map file, I can reproduce the bug.
I'll keep you posted.

**Comment by MKergall** [comment]
@monsieurtanuki No idea. I just picked an "adaptor" found on the Internet, and I packed it in OBP. 
Code of this adaptor is quite small - basically, just requesting MapsForge lib to build the tiles with relevant params. 
So, if it's not a MapsForge bug, it's going to be really tricky to understand and solve... 

@pavelkv96 You tested with MapsForge. But did you test with exactly the **same** version of MapsForge than the version used inside osmdroid? 
Current osmdroid is using org.mapsforge:mapsforge-map:0.8.0, which is quite old now.

**Comment by spyhunter99** [comment]
i can look into updating the adapter and probably cut a new version of osmdroid this weekend

**Comment by monsieurtanuki** [comment]
@pavelkv96 
> I tested with MapsForge and they don't have this bug

That's not that simple anyway: you experience the bug at zoom level 10, I do at level 14.
More something like a side-effect. About memory?
Maybe you would experience this bug with MapsForge but on a different spot and a different zoom level.

Just another test: if I go directly to the "suspicious" zone
````java
mMapView.getController().setZoom(14);
mMapView.setExpectedCenter(new GeoPoint(53.809, 23.709));
````
there's no label at all.
But, if I close the app, turn the Hardware Acceleration off (in "Settings"), and go back, then there's the label. Afterwards, zooming to 13 then to 14, only part of the label is displayed (the part on the right).

@pavelkv96 Could you try to run your app using that line:
````java
Configuration.getInstance().setMapViewHardwareAccelerated(false));
```` 

@spyhunter99 Perhaps the infamous "polyline/polygon rendering with hardware acceleration on" bug again...

**Comment by pavelkv96** [comment]
@monsieurtanuki 
May be bug with save tile? Because Mapsforge don't save their.
Because when I don't check tile in cache, and I generate anew, all work good

**Comment by monsieurtanuki** [comment]
@pavelkv96 What do you mean with "when I don't check tile in cache, and I generate a new"? How do you do that?

Besides that, unfortunately I don't think that your "save tile/don't save tile" hypothesis is right, because the first time the tile is displayed the label is correctly displayed.
For the record, the tile display works more or less that way:
* in the display loop, we try to find each tile in the _memory_ cache
* for each tile that is not in the memory cache, we loop asynchronously through all providers until one of them knows this tile, and then puts it in the memory cache and runs a display refresh

The extra step about the Mapsforge provider is that if it knows a tile, it has to compute and render it on a bitmap tile. For other providers it's more straight forward: I already have a bitmap, I give you this bitmap and there's no computation.

You say Mapsforge library seems to work fine in your case and there's obviously a bug in the Mapsforge x osmdroid combination: just out of curiosity, wouldn't you be better off using Mapsforge only, or are there special features in osmdroid that you need for your app? That doesn't mean the bug won't be fixed ;)

**Comment by MKergall** [comment]
> I tested with MapsForge and they don't have this bug

Which version of MapsForge?

**Comment by MKergall** [comment]
Looking at Mapsforge repo, I just discovered its VTM fork. I think it would be great to integrate VTM as a rendered of Mapsforge maps in osmdroid. 
With @devemux86 help?...

**Comment by devemux86** [comment]
Regarding VTM please see also my comment [here](https://github.com/osmdroid/osmdroid/issues/246#issuecomment-331084317).

**Comment by pavelkv96** [comment]
@MKergall version 0.8.0

**Comment by monsieurtanuki** [comment]
The problem comes from mapsforge. I've just created an issue (https://github.com/mapsforge/mapsforge/issues/1085).

Let me explain the problem.

When a tile is rendered, mapsforge looks for all the neighboring tiles' potential overlapping labels. That's cool.

But at some point, as we use this process in a slightly different way as planned (no direct display, not really using a mapsforge cache), there's a side effect about the label display.

In the mapsforge context, sometimes it makes sense to display the labels only once, even if they're overlapping on several tiles: displaying it once will cover both tiles.

In our context, we definitely always want to display the labels because we don't care about the neighboring tiles' bitmap being already processed or in cache: all we need to know is if they have overlapping labels.

More specifically, what happened is that the very first time we want to display "Василевичи" on level 14, both tiles are not processed yet and one of them says: "btw, display the left part of it, I'll display the right part". So far so good.
But the next time, after going to zoom 13 and back to 14, that's were the bug is in mapsforge - as both tiles have been processed (though we at osmdroid didn't necessarily cache them), mapsforge decides that only one of the tiles should display the label.

**Comment by monsieurtanuki** [comment]
The fix will be to use the new `DirectRenderer` class instead of `DatabaseRenderer` in `MapsForgeTileSource`.
Assuming that https://github.com/mapsforge/mapsforge/pull/1086 is merged and that we manage to upgrade our version of mapsforge to the latest.

**Comment by monsieurtanuki** [comment]
@pavelkv96 Could you try with #1152? Should be fixed.

**Comment by pavelkv96** [comment]
@monsieurtanuki , @MKergall , @spyhunter99 , @devemux86
Thanks to everyone for their help.

**Comment by monsieurtanuki** [comment]
@pavelkv96 It's not over yet, there are still some bugs, cf. https://github.com/mapsforge/mapsforge/issues/1085.

**Comment by secaw** [comment]
@monsieurtanuki Cool, this looks promising! The sometimes broken mapsforge labels have been annoying for a long time...

**Comment by monsieurtanuki** [comment]
@secaw Don't hesitate to provide me with samples of broken mapsforge labels. I need to test and test again.

**Comment by secaw** [comment]
@monsieurtanuki Unfortunately I noticed, that broken labels are not exactly reproducible across devices. I guess, it's dependend on the map scale (screen density), where it actually happens. As far as I saw, it occurs at some point in any map file I tried. For me it has improved after moving to mapsforge 0.9.1 (less occurences), so I haven't paid attention that much since then.

**Comment by monsieurtanuki** [comment]
Could be solved by #1156.

**Comment by secaw** [comment]
@monsieurtanuki Sounds very promising indeed. Does it recompute neighboring tiles always, or only under certain circumstances. I'm still very much aware of cumpute resources regarding mapsforge / osmdroid ongoing vector to pixel conversion.
Since you're PRing to mapsforge also, I have big wish: Would it be possible to return a NULL-tile, when requesting a tile that is out of bounds of the specific map file, like we do across any osmdroid providers? This would be great, so we can actually use mapsforge files as layers too. (just slightly related: #1072)
I found a solution for myself some time ago by customizing the osmdroid mapsforge implementation that way that I'm saving the bounds of the maps file and then checking against every tile, by converting it to its bounds coordinates. That is pretty computation intense on top of the actual tile rendering and might be solved way better, if mapsforge would just return by itself, if the map includes a tile or not (NULL).

**Comment by monsieurtanuki** [comment]
@secaw It recomputes only the neighboring tiles that are _already in the memory cache_ and that have a _new_ neighbor with overlapping labels.

Regarding the NULL-tile, I've just added this "out-of-box tile returns NULL" feature in Mapsforge's `DirectRenderer`. Obviously that doesn't solve everything, but this issue goes beyond my understanding of Mapsforge (something like bounding box vs. data range). You'll see the results.

**Comment by secaw** [comment]
@monsieurtanuki Fantastic, thank you very much. I'm looking forward to checking it all out...

**Comment by pavelkv96** [comment]
@monsieurtanuki How to solve this problem?

**Comment by monsieurtanuki** [comment]
@pavelkv96 We just have to wait the next mapsforge release, then I'll merge my osmdroid code.
Or are you talking about another issue than #1145?

**Comment by pavelkv96** [comment]
@monsieurtanuki Thank you. No, I'm talking about #1145
